### DISCUSSION TITLE
Instrumentation Allocations

### DISCUSSION THREAD

**Discussion by treverj** [graphql-java/graphql-java#2743] [Q&A]
First of all I want to thank you for a fantastic library which has made our move to providing a GraphQL based API immeasurably simpler. 

We have been investigating some performance problems with our API, which we suspect to be at least in part down to
prodigious object allocation within graphql-java.

The graphql-java Instrumentation framework allocates a new instance of InstrumentationXXXParameters for each subject it
visits, and for each Instrumentation registered when using ChainedInstrumentation. When using the field-level
instrumentation, these allocations contribute to a significant amount of memory pressure on the application when it is
under load. In our specific case, these newly-allocated objects aren't used at all, so the allocation is completely
wasteful.

The reason for the new allocation in ChainedInstrumentation is to pass along the correct state to the new subject. As a
proof-of-concept, we altered the signature of the Instrumentation::beginField, Instrumentation::beginFieldFetch and
Instrumentation::instrumentDataFetcher methods to accept both a InstrumentationFieldParameters and an
InstrumentationState, and removed the InstrumentationState member from InstrumentationFieldParameters. Along with code
changes to pass the state object along appropriately, this enabled us to avoid these problematic allocations.

Without changes: 4.48 TB of total allocations during a performance test
After changes: 3.61 TB of total allocations during a performance test

More specifically:

- 459GB of extra/unnecessary allocated memory (11.5bn objects) for InstrumentationFieldFetchParameters.withNewState
- 187GB of extra/unnecessary allocated memory (11.7bn objects) for IntstrumentationFieldFetchParameters intermediate lambda
- 136GB of extra/unnecessary allocated memory (5.65bn objects) for InstrumentationFieldParameters.withNewState

We realise that this dramatic and incompatible change to a key API is most likely not the appropriate solution to this
problem. As non-experts in both the graphql-java codebase and its design, would appreciate some assistance with
deriving a more appropriate design that might have a chance of being accepted as a PR.

Example PR: https://github.com/graphql-java/graphql-java/pull/2742

It is also entirely possible that we are mis-using the Instrumentation API here. Our primary motivations for using it
are that we are a multi-tenanted e-commerce platform provider, with some clients having bespoke features that only apply to them, and others having a wide array of features not enabled for their site. To simplify the schema significantly, we have created a new set of directives and instrumentation called SchemaHidingInstrumentation to enabled and disable the fields we want to hide from the schema. For introspections queries, this wraps the data fetcher to hide/remove fields from the schema and for queries or mutations that attempt to use these fields, we return a fixed "Field not available error". We also use this to remove fields we are trialing through experiments or gradual roll outs. 

If there's a more sensible way to achieve our goals here, we'd be happy to hear about it.

Thanks,
James

**Comment by enbnt** [comment]
More of a +1 than an answer, but we ran into a similar issue at Twitter, primarily around the copy/map behavior over each instrument within a `ChainedInstrumentation` in each of the `beginXXX` methods. We were able to do a workaround where we 

- extended from `DataLoaderDispatcherInstrumentation`
- override each of the `beginXXX` methods to iterate over each of the instrumentations and fire off the same method and return the result of calling `super.beginXXX`
-  and all of the `instrumentXXX` methods take a result, decorate it for each of the instrumentations, and pass that to the `super.instrumentXXX` methods. 
 
The catch to this is that this only works for instrumentations that do no state management (i.e. return a `SimpleInstrumentationContext.noOp()` where a `whenDispatched` or `whenCompleted` are never called). This method doesn't work with `TracingInstrumentation`, but it was critical for us to optimize this as we have scenarios where 45k+ fields are fetched.

It would be great to start a discussion around what would make sense from a graphql-java perspective. There would likely be some sort of breaking API change involved, but it would be great if we had the ability to differentiate between something that is side-effecting without state (i.e. update a metric) vs something that requires managed state via the `InstrumentationState` that requires heavier state management. We are happy to chat more about this.

Thanks,
Ian

**Reply by bbakerman** [reply]
Can you outline more about your work around?

I am confused mostly by the following and what you mean by `instrumentXXX` methods.  Maybe some pseudo code example might help me.

> and all of the instrumentXXX methods take a result, decorate it for each of the instrumentations, and pass that to the super.instrumentXXX methods.


Are you saying you chose one particular instrumentation (`DataLoaderDispatcherInstrumentation` which is always present) and called N custom instrumentations because you knew they never used complex `InstrumentationContext` return object?

> this only works for instrumentations that do no state management (i.e. return a SimpleInstrumentationContext.noOp() where a whenDispatched or whenCompleted are never called).

yeah so there is 2 aspects to the instrumentation state management.   There is the state object created via `graphql.execution.instrumentation.Instrumentation#createState(graphql.execution.instrumentation.parameters.InstrumentationCreateStateParameters)` and there is the afct you can create a call back `InstrumentationContext` object that will be called back when the async operation finishes

(a callback from a callback - meta! )

I guess to do what I think you are saying (inline without ever calling back on the individual `InstrumentationContext` objects we would need some marker interface (sa `StatelessInstrumentation` marker) and we adjust behavior base on that

**Reply by bbakerman** [reply]
> but it would be great if we had the ability to differentiate between something that is side-effecting without state (i.e. update a metric) 

Also just be clear here - there is probably state involved in a `metric` update because you have to run a `graphql.execution.instrumentation.InstrumentationContext#onCompleted` callback to say when the work is "done" say

(for timer metrics etc..)

**Reply by enbnt** [reply]
> Can you outline more about your work around?
> 
> I am confused mostly by the following and what you mean by `instrumentXXX` methods. Maybe some pseudo code example might help me.
> 

```
class FastChainedInstrumentation (
  instrumentations: Array[SimpleInstrumentation])
    extends DataLoaderDispatcherInstrumentation {

  /* Stateless methods that are only side-effecting and do not modify input */

  override def beginExecution(
    parameters: InstrumentationExecutionParameters
  ): InstrumentationContext[ExecutionResult] = {
    instrumentations.foreach(_.beginExecution(parameters))
    super.beginExecution(parameters)
  } 

  ...

  /* Stateful chained methods which decorate an input */

  override def instrumentExecutionContext(
    executionContext: ExecutionContext,
    parameters: InstrumentationExecutionParameters
  ): ExecutionContext = {
    var ec = executionContext
    instrumentations.foreach { inst =>
      ec = inst.instrumentExecutionContext(ec, parameters)
    }
    super.instrumentExecutionContext(ec, parameters)
  }
}
```

> > and all of the instrumentXXX methods take a result, decorate it for each of the instrumentations, and pass that to the super.instrumentXXX methods.
> 
> Are you saying you chose one particular instrumentation (`DataLoaderDispatcherInstrumentation` which is always present) and called N custom instrumentations because you knew they never used complex `InstrumentationContext` return object?
> 

This was specifically to avoid having a `ChainedInstrumentation` wrap our instrumentations in `GraphQL#checkInstrumentationDefaultState`, where it will not wrap as long as a `DataLoaderDispatcherInstrumentation` is present. Even a `ChainedInstrumentation` with just the default `DataLoaderDispatcherInstrumentation` and our `FastChainedInstrumentation` was a significant allocation source given the number of fields being processed.

> > this only works for instrumentations that do no state management (i.e. return a SimpleInstrumentationContext.noOp() where a whenDispatched or whenCompleted are never called).
> 
> yeah so there is 2 aspects to the instrumentation state management. There is the state object created via `graphql.execution.instrumentation.Instrumentation#createState(graphql.execution.instrumentation.parameters.InstrumentationCreateStateParameters)` and there is the afct you can create a call back `InstrumentationContext` object that will be called back when the async operation finishes
> 
> (a callback from a callback - meta! )
> 
> I guess to do what I think you are saying (inline without ever calling back on the individual `InstrumentationContext` objects we would need some marker interface (sa `StatelessInstrumentation` marker) and we adjust behavior base on that

Yes!


> > but it would be great if we had the ability to differentiate between something that is side-effecting without state (i.e. update a metric)
> 
> Also just be clear here - there is probably state involved in a `metric` update because you have to run a `graphql.execution.instrumentation.InstrumentationContext#onCompleted` callback to say when the work is "done" say
> 
> (for timer metrics etc..)

We typically have 5-6 different instrumentations being chained, where they are mostly incrementing a counter based on some state (not a per-invocation timing) or doing some sort of verification (check something via parameters, throw an exception if something if not valid). You are correct that today something that's timing-based would need onDispatched or onCompleted.

**Reply by bbakerman** [reply]
> as we have scenarios where 45k+ fields are fetched.

you have to give me insight into this - who do you have 45K+ fields in one query .   What drives that?

I ask out of morbid curiosity and a desire to see how graphql-java is getting used and hence how it can be improved

**Comment by bbakerman** [comment]
I will get back to you on this idea of passing state direct (rather than via a copied "parameters" object later.

But

> It is also entirely possible that we are mis-using the Instrumentation API here. Our primary motivations for using it
are that we are a multi-tenanted e-commerce platform provider, with some clients having bespoke features that only apply to them, and others having a wide array of features not enabled for their site. To simplify the schema significantly, we have created a new set of directives and instrumentation called SchemaHidingInstrumentation to enabled and disable the fields we want to hide from the schema. For introspections queries, this wraps the data fetcher to hide/remove fields from the schema and for queries or mutations that attempt to use these fields, we return a fixed "Field not available error". We also use this to remove fields we are trialing through experiments or gradual roll outs.

graphql-java has a two mechanisms for presenting different fields to consumers.

If you can pre categorize your consumers into static "sets" then you can use `graphql.schema.SchemaTransformer` and specifically `graphql.schema.transform.FieldVisibilitySchemaTransformation` to create new schema objects that have certain fields missing say.         This is typically done not per request but at say app startup time, since transforming a large schema can be computational expensive and hence not great per request.

There is also a  `graphql.schema.visibility.GraphqlFieldVisibility` that runs at request time.   This is an older mechanism (out first attempt) that decides if a field is visisble or not.  Again this is tied to the schema (via CodeRegistry) BUT its cheaper to `transform` schema to have one of these than it is to "rewrite" it structurally.  However it will be called every time we access a field and this can have its own "performance" challenges depending on what you do in this callback.

**Comment by bbakerman** [comment]
Ok I looked into this more and its makes sense.

The goal of the "new parameters with state" really so that the called instrumentation method could do `parameters.getState()` as a convenience AND also allow auto `<T`> casting eg `MyOwnState state = params.getState()`

But I can see how it makes sense to avoid object allocation for this at scale.  Of course this is ONLY a problem at scale.  Most systems would not notice this cost I think.

To confirm the breaking API change would be

* the call back methods themselves would have a second "state" parameter.  
* the `InstrumentationXXXXParameters` objects would no longer have the `getState()` method and `<T>` casting

Perhaps an idea to lessen impact is to only change the methods that get called a lot (field ones versus singleton ones like `graphql.execution.instrumentation.Instrumentation#beginParse` say

The other object allocation (that might be much harder to avoid) is the `ChainedInstrumentationContext` creation.

This delegates to N underlying `InstrumentationContext` instances because each call back method (might) need to know when it has been dispatched and mostly importantly when the async step has completed (`graphql.execution.instrumentation.InstrumentationContext#onCompleted`

This is tricky to avoid because you need hold each different one so it can be invoked.  Even if we had some "marker interface" say which said it a "simple" instrumentation call, you still need some allocation for ones that are not that.  And allocating a list of 2 is about the same cost as any other list.

**Comment by bbakerman** [comment]
@treverj / @enbnt   -for reference - who many instrumentations are you running here.  That gives the factor `N` allocations per field fetch

Also what instrumentations from graphql-java do you use and what ones are custom?

**Reply by treverj** [reply]
Hey @bbakerman, 

Thanks for looking into this and the discussion/thoughts around this.

We currently run 15 different pieces of instrumentation. At a high level, here are what they do:
- Dynamic Type Instrumentation (adds some custom objects dynamically at run time for end user defined objects)
  - Uses instrumentSchema method
  - Doesn't rely on onDispatched/onCompleted
- Distributed tracing instrumentation (adds tracing info to response) 
  - Uses instrumentExecutionResult
  - Doesn't rely on onDispatched/onCompleted
- Feature flag instrumentation (adds a few fields to the extension around currently active feature flags)
  - Uses instrumentExecutionResult
  - Doesn't rely on onDispatched/onCompleted
- Experiment Instrumentation (adds AB testing info to the extension based on the specific user making the request)
  - Uses instrumentExecutionResult
  - Doesn't rely on onDispatched/onCompleted
- Tracking Instrumentation (this perfroms some async analytics about the requests we are receiving for data analysis)
  - Uses beginParse 
  - Doesn't rely on onDispatched/onCompleted - returns `SimpleInstrumentationContext.noOp()`
- Schema HidingInstrumentation (this is the one we see issues with as it checks every field)
  - Uses instrumentDataFetcher
  - Doesn't rely on onDispatched/onCompleted
- DescriptionIntrospectionFixInstrumentation - temp fix for https://github.com/graphql-java/graphql-java/issues/2410, this can be removed now
  - Uses instrumentDataFetcher
  - Doesn't rely on onDispatched/onCompleted
- Maximum Weight Instrumentation - adds weights to specific queries and prevents queries that are too expensive being used together
  - Uses beginValidation and instrumentExecutionResult
  - Relies on **onCompleted**
- MaxQueryDepthInstrumentation - from graphql java
- Rate Limiting Instrumentation
  - Uses instrumentExecutionResult and instrumentDataFetcher (to determine if a given query/mutation has rate limiting for it)
  - Doesn't rely on onDispatched/onCompleted
- Require Authentication Instrumentation - validations if a given field/query/mutation requires the user to be authenticated
  - Uses instrumentDataFetcher
  - Doesn't rely on onDispatched/onCompleted
- Volatile Instrumentation - marks a given field as volatile and could be changed later for experimental new changes
  - Uses beginField and instrumentExecutionResult
  - Doesn't rely on onDispatched/onCompleted - adds a field to the instrumentation state in the `InstrumentationFieldParameters` and returns `super.beginField(parameters);` in the beginField method
- ResponseTimingInstrumentation - times how long requests take
  - Uses beginExecution
  - Uses **onCompleted**
- Access Logging Instrumentation - does the equivalent of standard access logs but for graphql, outputting the query and request information, removing sensitives information
  - Uses createState, instrumentDocumentAndVariables, beginExecutionStrategy and instrumentExecutionResult
  - Doesn't rely on onDispatched/onCompleted
- Metrics Recording Instrumentation - logs metrics and timings for different stages of graphql operations
  - Uses beginExecution, beginParse, beginValidation, beginFieldFetch (with customer contexts used to finish the metrics)
  - Doesn't rely on onDispatched/onCompleted

**Reply by bbakerman** [reply]
Thanks for the reply.  Can you expand on this list (edit the reply maybe) and indicate whether you ever rely on

`graphql.execution.instrumentation.InstrumentationContext#onDispatched` or `graphql.execution.instrumentation.InstrumentationContext#onCompleted` running 

The reason is this.  One idea is to create a "fast instrumentation" method call.  Chaining can be simplified if you never need to returned context to call you back.

It can be just a series of call one after the other (discard the need to call back to the context say if an Instrumentation is marked as `NoContexNeeded` say (future marker interface)

The other way to do that of course is to create a `NoContextChainedInstrumentation` class say that contains these type of cases and hence it fit inside a Chained INstrumentation say and we require the consumers to wire it together themselves

**Reply by treverj** [reply]
Updated, we only have the ResponseTime and MaxWeigh Instrumentation that relies on the InstrumentationContext for those 2 methods so your suggestion sounds like but would work in our case for most of our instrumentation.

**Comment by bbakerman** [comment]
I created a Spike PR (which is pretty much the PR that was submitted but tweaked) to show how we could make this change in a non API breaking way

https://github.com/graphql-java/graphql-java/pull/2747

**Reply by treverj** [reply]
Is this something that would be possible to get merged in? Or are you thinking of using something more akin to the `NoContextChainedInstrumentation` or `FastChainedInstrumentation` mentioned above?

**Reply by bbakerman** [reply]
This maybe be merged in at some future time - we are trying to get v18 right now so it wont be soon

As a work around (and perhaps as a permenant strategy) you would be better with  specialised FastChainedInstrumentation instead since you are kinda special (super large scale - TB of memory) 

I think any future PR would be to "pass state" direct in but it would always run Context callbacks (and hence have to allocate lists to hold them even if they are empty

**Comment by bbakerman** [comment]
@treverj @enbnt 

I had this idea recently if we added new Instrumentation methods (and deprecated the old)

re: from https://github.com/graphql-java/graphql-java/pull/2747

The new shape would be this

```java
 default @Nulllable InstrumentationContext<Object> beginFieldFetch(InstrumentationFieldFetchParameters parameters, InstrumentationState state) {
       // calls back to old behavior always
        return beginFieldFetch(parameters.withNewState(state));
    }
```

That is the state gets passed in (and hence no object allocation for changing the "parameters" and the return `InstrumentationContext` is now allowed to be `null`.

So instrumentations that don't want any callback would return null, also avoiding an object allocation.  They can still be called in a chain (which would be very much like your `FastChainedInstrumentation` idea) and those instrumwentations that are stateless in all the ways can return null and hence no object allocation

The impact on our code would be we need to "handle" null in all the places that call back but I think this would be ok. (We can do a trick like `InstrumentationContext<T> ic = returnedIC == null ? staticNoAllocaNoOp : returnedIC`
